cpmpy/expressions/globalconstraints.py

```diff
@@ -172,7 +172,10 @@ class AllDifferent(GlobalConstraint):
     """All arguments have a different (distinct) value
     """
     def __init__(self, *args):
-        super().__init__("alldifferent", flatlist(args))
+        flatargs = flatlist(args)
+        if not (all(is_boolexpr(arg) for arg in flatargs) or not any(is_boolexpr(arg) for arg in flatargs)):
```

Comment:
again, flatlist is not calling flatten, or recursively giving back all children, it just unnests iterables and gives back a list. (since the Alldifferent constructor allows things like AllDifferent([1,2,[1,3]]) and is_ boolexpr on a list will always return false so doing [is_boolexpr(a) for a in args] would run into issues that way

Reply:
but certainly it could be more efficient than looping twice, maybe a custom function that checks the type of the first element and then checks if the other elements are of the same type?